clang/lib/CIR/Dialect/Transforms/LoweringPrepareAArch64CXXABI.cpp

```diff
@@ -0,0 +1,258 @@
+  // argument. We don't want to keep updating regOffs (in case it overflows,
+  // though anyone passing 2GB of arguments, each at most 16 bytes, deserves
+  // whatever they get).
+  auto zeroValue = builder.create<mlir::cir::ConstantOp>(
+      loc, regOffs.getType(), mlir::cir::IntAttr::get(regOffs.getType(), 0));
+  auto usingStack = builder.create<mlir::cir::CmpOp>(
+      loc, boolTy, mlir::cir::CmpOpKind::ge, regOffs, zeroValue);
+  builder.create<mlir::cir::BrCondOp>(loc, usingStack, onStackBlock,
+                                      maybeRegBlock);
+
+  auto contBlock = currentBlock->splitBlock(op);
+
+  // Otherwise, at least some kind of argument could go in these registers, the
+  // question is whether this particular type is too big.
+  builder.setInsertionPointToEnd(maybeRegBlock);
+
+  // Integer arguments may need to correct register alignment (for example a
+  // "struct { __int128 a; };" gets passed in x_2N, x_{2N+1}). In this case we
+  // align __gr_offs to calculate the potential address.
+  if (!IsFPR) {
+    assert(!cir::MissingFeatures::handleAArch64Indirect());
+    assert(!cir::MissingFeatures::supportTyAlignQueryForAArch64());
+  }
+
+  // Update the gr_offs/vr_offs pointer for next call to va_arg on this va_list.
+  // The fact that this is done unconditionally reflects the fact that
+  // allocating an argument to the stack also uses up all the remaining
+  // registers of the appropriate kind.
+  auto regSizeValue = builder.create<mlir::cir::ConstantOp>(
+      loc, regOffs.getType(),
+      mlir::cir::IntAttr::get(regOffs.getType(), regSize));
+  auto newOffset = builder.create<mlir::cir::BinOp>(
+      loc, regOffs.getType(), mlir::cir::BinOpKind::Add, regOffs, regSizeValue);
+  builder.createStore(loc, newOffset, regOffsP);
+  // Now we're in a position to decide whether this argument really was in
+  // registers or not.
+  auto inRegs = builder.create<mlir::cir::CmpOp>(
+      loc, boolTy, mlir::cir::CmpOpKind::le, newOffset, zeroValue);
+  builder.create<mlir::cir::BrCondOp>(loc, inRegs, inRegBlock, onStackBlock);
+
+  //=======================================
+  // Argument was in registers
+  //=======================================
+  // Now we emit the code for if the argument was originally passed in
+  // registers. First start the appropriate block:
+  builder.setInsertionPointToEnd(inRegBlock);
+  auto regTopP = builder.createGetMemberOp(
+      loc, valist, IsFPR ? "vr_top" : "gr_top", regTopIndex);
+  auto regTop = builder.create<mlir::cir::LoadOp>(loc, regTopP);
+  auto i8Ty = mlir::IntegerType::get(builder.getContext(), 8);
+  auto i8PtrTy = mlir::cir::PointerType::get(builder.getContext(), i8Ty);
+  auto castRegTop = builder.createBitcast(regTop, i8PtrTy);
+  // On big-endian platforms, the value will be right-aligned in its stack slot.
+  // and we also need to think about other ABI lowering concerns listed below.
+  assert(!cir::MissingFeatures::handleBigEndian());
```

Comment:
We should go a bit further here. The skeleton is really important:

```
// TODO(cir): implement check for jomogeneous aggregate
assert(!cir::MissingFeatures::supportisHomogeneousAggregateQueryForAArch64());
...
uint64_t NumMembers = <expected-default>...
bool IsHFA = <expected-default>...
if (IsHFA && NumMembers > 1) {
    ... // or unrecheable
}  else {
    ... // or unrecheable
}
```
For the path that it's not unrecheable, add more skeleton and perhaps more unrecheable for other inner/sub checks that are not implemented/tested.